### PR TITLE
feat: support TS directory imports for ESM

### LINKED ISSUE #421: Can't use typescript commands with ESM modules
Hi there,

I'm trying to create an oclif CLI as an ESM package that uses typescript, but am having trouble with using commands with the `.ts` extension. The CLI works fine if my command files are `.js` files, but the minute I rename them to `.ts` I get the following error:

```sh
$ ./bin/dev.js 
(node:73944) [ERR_REQUIRE_ESM] Error Plugin: my-cli: Must use import to load ES Module: /cli/src/commands/create/index.ts
require() of ES modules is not supported.
require() of /cli/src/commands/create/index.ts from /cli/node_modules/@oclif/core/lib/module-loader.js is an ES module file as it is a .ts file whose nearest parent package.json contains "type": "module" which defines all .ts files in that package scope as ES modules.
Instead change the requiring code to use import(), or remove "type": "module" from /cli/package.json.

module: @oclif/core@1.8.2
task: toCached
plugin: my-cli
root: /cli
See more details with DEBUG=*
(Use `node --trace-warnings ...` to show where the warning was created)
The CLI

VERSION
  my-cli/2.0.0 darwin-x64 node-v16.15.0

USAGE
  $ cli [COMMAND]

TOPICS
  plugins  List installed plugins.

COMMANDS
  help     Display help for cli.
  plugins  List installed plugins.
```

I tried digging around in the module-loader doing things like adding the `.ts` extension to `s_EXTENSIONS` and the `isPathModule()` method, but only seem to be digging a deeper hole. Does anyone have an example of how this is expected to work?

### ./bin/dev.js

```js
#!/usr/bin/env node

import oclif from '@oclif/core';
import path from 'node:path';
import url from 'node:url';
import { register } from 'ts-node';

const project = path.join(path.dirname(url.fileURLToPath(import.meta.url)), '..', 'tsconfig.json');

// In dev mode -> use ts-node and dev plugins
process.env.NODE_ENV = 'development';

register({ project });

// In dev mode, always show stack traces
oclif.settings.debug = true;

// Start the CLI
oclif
  .run(void 0, import.meta.url)
  .then(oclif.flush)
  .catch(oclif.Errors.handle);
```

### package.json

```json
{
  "name": "my-cli",
  "version": "2.0.0",
  "description": "The CLI",
  "type": "module",
  "exports": {
    ".": "./dist/index.js"
  },
  "files": [
    "/bin",
    "/dist",
    "/npm-shrinkwrap.json",
    "/oclif.manifest.json"
  ],
  "bin": {
    "cli": "./bin/run"
  },
  "oclif": {
    "bin": "cli",
    "dirname": "cli",
    "commands": "./dist/commands",
    "plugins": [
      "@oclif/plugin-help",
      "@oclif/plugin-plugins"
    ],
    "topicSeparator": " "
  },
  "dependencies": {
    "@oclif/core": "^1.8.2",
    "@oclif/plugin-help": "^5.1.12",
    "@oclif/plugin-plugins": "^2.1.0",
    "ajv": "^8.11.0",
    "axios": "^0.27.2",
    "cli-table": "^0.3.11",
    "execa": "^6.1.0",
    "inquirer": "^8.2.4"
  },
  "devDependencies": {
    "@oclif/test": "^2.1.0",
    "@types/cli-table": "^0.3.0",
    "@types/inquirer": "^8.2.1",
    "prettier": "^2.6.2",
    "shx": "^0.3.4",
    "ts-node": "^10.7.0",
    "typescript": "^4.6.4"
  },
  "scripts": {
    "build": "shx rm -rf dist && tsc -b",
    "lint": "eslint . --ext .ts --config .eslintrc",
    "postpack": "shx rm -f oclif.manifest.json",
    "posttest": "npm run lint",
    "prepack": "npm run build && oclif manifest && oclif readme",
    "version": "oclif readme && git add README.md"
  }
}
```

### ./src/commands/create/index.ts

```typescript
import { Command } from '@oclif/core';

export default class CreateCommand extends Command {
  static description = 'Create a new resource';

  async run() {}
}
```

### PR DESCRIPTION
**Pull request by mdonnalley**
Support ESM module resolution for typescript directories

Fixes #421

### CODE REVIEW

File: src/module-loader.ts [hunk 1]
```diff
@@ -13,7 +13,7 @@ const getPackageType = require('get-package-type')
  * Defines file extension resolution when source files do not have an extension.
  */
 // eslint-disable-next-line camelcase
-const s_EXTENSIONS: string[] = ['.js', '.mjs', '.cjs']
+const s_EXTENSIONS: string[] = ['.ts', '.js', '.mjs', '.cjs']
```
Comment (peternhale) [inline src/module-loader.ts:16]: Should `.jsx` or `.tsx` file extensions be included?

**Review by peternhale — CHANGES_REQUESTED** [review]

File: src/module-loader.ts (hunk 1 above)
Comment (peternhale) [reply src/module-loader.ts:16]: Yes. Anything in oclif that prevents one from using react?

Comment (mdonnalley) [reply src/module-loader.ts:16]: @peternhale I just don't see the use case for using jsx syntax in a CLI. I'd rather not support it until someone comes along with a convincing use case

Comment (RodEsp) [reply src/module-loader.ts:16]: I definitely want to play around with https://github.com/vadimdemedes/ink at some point. Might add it then 😄

**Review by peternhale — APPROVED** [review]

**Review by RodEsp — APPROVED** [review]

**Pull request merged**